### PR TITLE
Feature/kafka upgrade

### PR DESCRIPTION
**Pull request by plirglo**
Created role for Kafka upgrade to version 2.3.1.

### CODE REVIEW

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/update-properties.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+---
+- name: Is server.properties file exists
+  stat:
+    path: "{{ specification.kafka_var.conf_dir }}/server.properties"
+
+- name: Is defined current_kafka_version property
+  shell: grep "^CURRENT_KAFKA_VERSION" {{ specification.kafka_var.conf_dir }}/server.properties
+  register: current_kafka_version_property
+  ignore_errors: yes
+
```
Comment (to-bar) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/update-properties.yml]:
```suggestion
  changed_when: false

```

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/update-properties.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+- name: Add current_kafka_version property
+  lineinfile:
+    path: "{{ specification.kafka_var.conf_dir }}/server.properties"
+    line: "CURRENT_KAFKA_VERSION={{ specification.kafka_var.version }}"
+  when: current_kafka_version_property.stdout == ""
+
+- name: Is defined inter.broker.protocol.version property
```
Comment (to-bar) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/update-properties.yml]:
```suggestion
- name: Check if inter.broker.protocol.version property is defined
```

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/update-properties.yml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  shell: grep "^inter.broker.protocol.version" {{ specification.kafka_var.conf_dir }}/server.properties
+  register: inter_broker_protocol_version_property
+  ignore_errors: yes
```
Comment (to-bar) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/update-properties.yml]: Please do not use `ignore_errors` but `failed_when` with `rc` or `stdout` or `stderr` based condition

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/update-properties.yml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
```
Comment (to-bar) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/update-properties.yml]:
```suggestion
  changed_when: false

```

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/update-properties.yml (hunk 3 above)
Comment (plirglo) [reply core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/update-properties.yml]: Added:

  failed_when:
    - result.rc == 2

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka.yml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
```
[the remark this answers is not in the document]
Comment (plirglo) [reply core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka.yml]: - names added
- created pre-check playbook where check version of actual kafka installation and skip if installed version is not older than target one.

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/set-updated-version.yml [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    path: /opt/kafka/config/server.properties
+  changed_when: False
+
+- name: Modify inter.broker.protocol.version property
+  lineinfile:
+    path: /opt/kafka/config/server.properties
+    regexp: "^inter.broker.protocol.version"
+    line: "inter.broker.protocol.version={{ specification.kafka_var.version }}"
+
+- name: Restart kafka service
```
Comment (to-bar) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/set-updated-version.yml]:
```suggestion
- name: Start kafka service
```

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/update-properties.yml [hunk 7]
```diff
@@ -0,0 +1,34 @@
+---
+- name: Check if kafka_version_preference property is defined
```
Comment (to-bar) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/update-properties.yml]:
```suggestion
- name: Check if current_kafka_version property is defined 
```

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/update-properties.yml [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  failed_when:
+    - result.rc == 2
+  changed_when: False
+  args: 
+    executable: /bin/bash
+
+- name: Add current_kafka_version property
+  lineinfile:
+    path: /opt/kafka/config/server.properties
+    line: "CURRENT_KAFKA_VERSION={{ before_upgrade_kafka_version.stdout }}"
```
Comment (to-bar) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/update-properties.yml]:
```suggestion
    line: "CURRENT_KAFKA_VERSION={{ before_upgrade_kafka_version }}"
```

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/update-properties.yml [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  failed_when:
+    - result.rc == 2
+  changed_when: False
+  args:
+    executable: /bin/bash
+
+- name: Add inter.broker.protocol.version property
+  lineinfile:
+    path: /opt/kafka/config/server.properties
+    line: "inter.broker.protocol.version={{ before_upgrade_kafka_version.stdout }}"
```
Comment (to-bar) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/update-properties.yml]:
```suggestion
    line: "inter.broker.protocol.version={{ before_upgrade_kafka_version }}"
```

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka.yml [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    include_tasks: kafka/update-properties.yml
+
+  - name: Include kafka upgrade task 
+    include_tasks: kafka/install-upgrade.yml
+
+  - name: Include kafka version set
+    include_tasks: kafka/set-updated-version.yml
+  when:
+    - before_upgrade_kafka_version.stdout is defined
+    - lock_file_status.stat.exists or before_upgrade_kafka_version.stdout is version( specification.kafka_var.version , '<' )
```
Comment (to-bar) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka.yml]:
```suggestion
    - lock_file_status.stat.exists or before_upgrade_kafka_version is version( specification.kafka_var.version , '<' )
```

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/install-upgrade.yml [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+- name: Link /opt/kafka to recently installed version
+  file:
+    dest: /opt/kafka
+    state: link
+    src: /opt/kafka_{{ specification.kafka_var.scala.version }}-{{ specification.kafka_var.version }}
+    force: yes
+
+- name: Remove previous version binaries
+  file:
+    path: /opt/kafka_{{ specification.kafka_var.scala.version }}-{{ before_upgrade_kafka_version.stdout }}
```
Comment (to-bar) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kafka/install-upgrade.yml]:
```suggestion
    path: /opt/kafka_{{ specification.kafka_var.scala.version }}-{{ before_upgrade_kafka_version }}
```

File: core/src/epicli/data/common/ansible/playbooks/upgrade.yml [hunk 12]
```diff
@@ -60,4 +60,14 @@
...
+
+- hosts: kafka
+  serial: 1
+  become: true
+  become_method: sudo
+  tasks:
+  - import_role:
+      name: upgrade
+      tasks_from: kafka
+    vars: {lock_file: /var/tmp/upgrade-kafka-in-progress.flag}
```
Comment (to-bar) [inline core/src/epicli/data/common/ansible/playbooks/upgrade.yml]:
```suggestion
    vars: {lock_file: /var/tmp/kafka-upgrade-in-progress.flag}
```

**Pull request merged**